src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -0,0 +1,571 @@
+    for _ in range(conf.workers):
+        conf.input_queue.put(None)
+
+    results = [conf.output_queue.get() for _ in range(assembly_files_len)]
+    assembly_name_to_upa = {assembly_name: upa for assembly_name, upa in results if upa is not None}
+    failed_names = [assembly_name for assembly_name, upa in results if upa is None]
+
+    # Close and join the processes
+    conf.pools.close()
+    conf.pools.join()
+
+    return assembly_name_to_upa, failed_names
+
+
+def main():
+    parser = _get_parser()
+    args = parser.parse_args()
+
+    workspace_id = args.workspace_id
+    kbase_collection = getattr(args, loader_common_names.KBASE_COLLECTION_ARG_NAME)
+    source_version = getattr(args, loader_common_names.SOURCE_VER_ARG_NAME)
+    root_dir = args.root_dir
+    token_filepath = args.token_filepath
+    upload_file_ext = args.upload_file_ext
+    workers = args.workers
+    keep_job_dir = args.keep_job_dir
+
+    env = args.env
+    kb_base_url = loader_common_names.KB_BASE_URL_MAP[env]
+
+    if workspace_id <= 0:
+        parser.error(f"workspace_id needs to be > 0")
+    if workers < 1 or workers > cpu_count():
+        parser.error(f"minimum worker is 1 and maximum worker is {cpu_count()}")
+
+    uid = os.getuid()
+    username = os.getlogin()
+
+    job_dir = loader_helper.make_job_dir(root_dir, username)
+    collection_source_dir = loader_helper.make_collection_source_dir(
+        root_dir, loader_common_names.DEFAULT_ENV, kbase_collection, source_version
+    )
+    upload_dir = loader_helper.make_collection_source_dir(root_dir, env, kbase_collection, source_version)
+    output_dir = loader_helper.make_sourcedata_ws_dir(root_dir, env, workspace_id)
+
+    proc = None
+    conf = None
+
+    try:
+        # start podman service
+        proc = loader_helper.start_podman_service(uid)
+
+        # set up conf, start callback server, and upload assemblies to workspace
+        conf = Conf(job_dir, kb_base_url, token_filepath, workers)
+        workspace_name = conf.ws.get_workspace_info({"id": workspace_id})[1]
+
+        count, wait_to_upload_assemblies = _fetch_assemblies_to_upload(
+            env, 
+            workspace_id, 
+            collection_source_dir,
+            upload_file_ext,
+        )
+
+        if not wait_to_upload_assemblies:
+            print(f"All {count} assembly files already exist in workspace {workspace_id}")
+            return
+
+        wtus_len = len(wait_to_upload_assemblies)
+        print(f"Originally planned to upload {count} assembly files")
+        print(f"Detected {count - wtus_len} assembly files already exist in workspace")
+
+        start = time.time()
+        data_dir = _prepare_skd_job_dir_to_upload(job_dir, wait_to_upload_assemblies)
+        assembly_name_to_upa, failed_names = upload_assembly_files_in_parallel(conf, workspace_name, os.listdir(data_dir))
+
+        if failed_names:
+            print(f"\nFailed to upload {failed_names}")
```

Comment:
If any failed I think the script should fail. Otherwise the user may not notice the failure note in the logs

Reply:
I suspect we would not be able to upload ~30000 GTDB assembly files into target WS in one shot. Many iterations will be needed. Therefore, IMO the script should only record the failure, but not necessarily to halt or fail.

Reply:
> I suspect we would not be able to upload ~30000 GTDB assembly files into target WS in one shot

That is absolutely the goal here. Eventually the uploader will be expected to run unattended, triggered by some automated system.

(also it's > 300K files so even worse)

Automation is not in scope today, but making the script robust enough to catch and recover from failures is. Part of that is not returning a successful error code when some uploads have failed, as well as making it very clear in the final output that  some uploads have failed. That allows for automated failure tracking in the future so we can roll up failures and figure out which ones we need to fix first

> Therefore, IMO the script should only record the failure, but not necessarily halt or fail.

At this point it's already done all the uploads that didn't fail, so I don't understand why failing is an issue?

Reply:
Based on this [comment](https://github.com/kbase/collections/pull/381#discussion_r1283828341), we probably end up putting [these lines](https://github.com/kbase/collections/blob/e04a6d3ed13106cf8fd91089bc7a86abb30a88de/src/loaders/workspace_uploader/workspace_uploader.py#L547-L555) in _process_input() and failure note will be moved to the end of the script. So not sure what we discuss here is relevant anymore.

Reply:
I think the remaining issue is that if failures occur, the script just prints out some text in the middle of the output and returns a successful error code. The script should either throw an exception or print out failures at the very end and return a failing error code.

Reply:
Updated!

Reply:
:+1: